GIT-3: Use bare cloning

Haxsen opened the pull request:
[Issue link](https://jira.xwiki.org/browse/GIT-3)
Hi, I know this PR is late as I was catching up with my university assignments.

Changes are explained in commit messages.
Coverage is 0.91+.
Extension built and tested on XWiki build.

Summary:
Added `getRepositoryBare` to get and save the repositories as Bare in local drive.
1. Introduce `getRepositoryBare` method in the interface and implement it in `DefaultGitManager`.
2. Introduce 2 `getRepositoryBare`, 1 for authentication, 2 for without authentication in `GitScriptService`.
3. Test bare repository by checking WorkTree in `DefaultGitManagerTest`.
4. Test use cases by counting commits and finding authors in `GitScriptServiceTest`.

Thanks.

Haxsen committed:
Add getRepositoryBare for saving bare repositories

Haxsen committed:
Make a common getGitRepository method and add getRepositoryBare

Explanation: getRepositoryBare method will only have 1 line added, i.e. cloneCommand.setBare (true/false)
So it's better to have a single private method to keep code efficiency.
Also some minor fixes and warnings are removed.

Haxsen committed:
Add getRepositoryBare for both public and private repositories

Notice: For the authentication-less getRepositoryBare method, the same getRepositoryBare is used but with null credentials which are handled by DefaultGitManager.
This will keep the code efficiency and we won't need extra method in Interface to be implemented.

Haxsen committed:
Add test for getRepositoryBare

Explanation: Since we only need to test the Bare repository and not the authentication (which has already been tested), so the credentials are provided valid hence the repository can be accessed by the new test.
getRepositoryBareWithCredentials() tests whether the resulting repository contains WorkTree. Since bare repositories have no WorkTree, a NoWorkTree exception is thrown, which means our test for bare repository has passed.

Haxsen committed:
Add getRepositoryBare tests

Explanation as follows.
getRepositoryBareAndCountCommits: Tests the getRepositoryBare method without authentication and counts commits then compare.
getRepositoryBareWithCredentialsAndFindAuthors: Tests the getRepositoryBare method with authentication (to keep coverage > 0.91) and finds authors then compares the result.

vmassol commented on src/main/java/org/xwiki/git/GitManager.java:
```diff
@@ -68,6 +68,25 @@ default Repository getRepository(String repositoryURI, String localDirectoryName
     }
 
+    /**
+     * Clone a private Git repository as Bare using the credentials provided by user and store it locally in the
```
I don't think this is true since it should be possible to clone as bare even with no credentials. 

I think we're also missing a method for that BTW.

Haxsen replied:
Changed.

vmassol commented on src/main/java/org/xwiki/git/GitManager.java:
```diff
@@ -68,6 +68,25 @@ default Repository getRepository(String repositoryURI, String localDirectoryName
+     * @param repositoryURI the URI to the Git repository to clone (eg "git://github.com/xwiki/xwiki-commons.git")
+     * @param localDirectoryName the name of the directory where the Git repository will be cloned (this directory is
+     *        relative to the permanent directory
+     * @param username the username of the Git user
```
not precise enough, doesn't tell that you can pass null. I think it could be best to use Optional.

vmassol replied:
Same for the previous commit/PR, we might want to use Optional.

vmassol replied:
Actually I think it could be best to pass a CredentialsProvider object instead. Offers more possibilities (like not just username/password), it's more generic.

vmassol replied:
ofc the script service API would still use strings but I'm referring here to the Java API.

vmassol replied:
For the credentials and bare use case, I'm actually wondering if we shouldn't allow passing a CloneCommand object for complex use cases. This opens up all possibilities of configuring it the way the caller needs it. We would just call the following when the passed clonecommand object is not null:

```
setDirectory(localDirectory)
                    .setURI(repositoryURI)
                    .call();
```

(if null then we build a basic one ourselves)

WDYT?

The rationale is that I don't see the point in recreating the jgit API... :)The added value that we bring is the XWiki integration, i.e. the storage location and clone there.

Haxsen replied:
CredentialsProvider seems like a better option, I am doing the changes.
About complex use cases, shouldn't the caller use his own implementation since this is only about getting the repository and saving it locally?
We can pass CloneCommand object though, but I believe it's better to make a new method for it. WDYT?

vmassol replied:
For the script API sure. For the java API I'm less sure and I think I'd go with passing a CloneCommand. If you look at our implementation you'll find that the clonecommand part is only these few lines and they don't bring much value: https://github.com/xwiki-contrib/api-git/blob/a6e7cf8d40595fed34643029eefac1162d11a757/src/main/java/org/xwiki/git/internal/DefaultGitManager.java#L94-L99 

The value is the rest: the location and the git init. The CloneCommand is just configuration and is the same as passing parameters except that you're nowhere need being able to specify all your can set on a clone command (and what about when jgit adds support for a new config option in CloneCommand)?

So for me passing a CloneCommand is a way to be able to pass the configuration, in a generic and future-proof way. 

Also, my API-warnings start to blink when you have API methods with more than 3-4 parameters, it's a sign that there's a design problem. We're at 5 now and we just need to look up what else can be configured and we'll get to 7-8 in no time...

vmassol replied:
> and we'll get to 7-8 in no time...

setFs(), setRemote(), setBranch(), setCloneAllBranches(), setMirror(), setCloneSubmodules(), setBranchesToClone(), setTagOption(), setTagOption(), setNoCheckout(), etc

Haxsen replied:
True, I now replaced them by CloneCommand.

vmassol commented on src/main/java/org/xwiki/git/GitManager.java:
```diff
@@ -68,6 +68,25 @@ default Repository getRepository(String repositoryURI, String localDirectoryName
+     * @param localDirectoryName the name of the directory where the Git repository will be cloned (this directory is
+     *        relative to the permanent directory
+     * @param username the username of the Git user
+     * @param accessCode the password or OAuth or personal access token that authenticates with the Git user
```
Same comment

Haxsen replied:
Updated.

vmassol commented on src/main/java/org/xwiki/git/internal/DefaultGitManager.java:
```diff
@@ -97,13 +91,16 @@ public Repository getRepository(String repositoryURI, String localDirectoryName,
             if (!gitDirectory.exists()) {
-                CloneCommand cloneCommand = git.cloneRepository();
+                CloneCommand cloneCommand = Git.cloneRepository();
+                // Bare becomes true if useBare is received true, otherwise false.
```
I don't find that comment very useful, do you? :)

Haxsen replied:
Nope. Let me remove it.

vmassol commented on src/test/java/org/xwiki/git/internal/DefaultGitManagerTest.java:
```diff
@@ -109,6 +111,23 @@ public void getRepositoryWhenWrongCredentialsProvided() throws Exception
     }
 
+    @Test
+    public void getRepositoryBareWithCredentials() throws Exception
```
There's no need to test with credentials again since it's been tested already in another test.

Haxsen replied:
Updated.

vmassol commented on src/test/java/org/xwiki/git/internal/DefaultGitManagerTest.java:
```diff
@@ -109,6 +111,23 @@ public void getRepositoryWhenWrongCredentialsProvided() throws Exception
+            fail("Expected NoWorkTreeException");
+        } catch (NoWorkTreeException expected) {
+            assertTrue(ExceptionUtils.getRootCauseMessage(expected).matches("NoWorkTreeException: .*"));
+        }
```
This is missing some explanation. Can you explain in a comment how this verifies that the repo is bare?

Haxsen replied:
Bare repository do not has a Work tree. So when we call `repository.getWorkTree()` an exception is thrown, i.e. `NoWorkTreeException`. If such exception is thrown, the repository has been tested to be Bare instead of non-bare.
If there is a WorkTree, the test fails because the repository is non-bare.

vmassol replied:
yes but don't explain it to me ;) Explain it to the rest of the world in the code. Now isn't there a way to get if a repo is a bare one instead (since this is an indirect solution)?

Haxsen replied:
Actually I thought `this` comment not the code comment.
I have replaced this method with `isBare()` now.

vmassol commented on src/test/java/org/xwiki/git/script/GitScriptServiceTest.java:
```diff
@@ -114,4 +117,35 @@ public void getCountCommits() throws Exception
+            Set<PersonIdent> authors = service.findAuthors(repository);
+            assertEquals(1, authors.size());
+            assertEquals("test author", authors.iterator().next().getName());
+        }
```
I have the feeling there's a lot of duplication in tests here over the existing tests.

Haxsen replied:
Minimized the duplication now.

9inpachi replied:
@Haxsen I think the duplication of use cases exists in your updated implementation.
For example, the use cases of counting commits and finding authors are being tested twice both for bare repository and a normal one.
Just a single test that simply tests if a repository is bare should be enough for testing bare repository and covering the code. The other use cases are already covered. Or better, you can test a use case unique to a bare repository.

Haxsen replied:
@9inpachi  I see, if that is what it means then let me update this PR in an hour.
Thanks a lot.

Haxsen replied:
Hi @9inpachi , I have fixed this, can you confirm? Thanks.

9inpachi replied:
Great @Haxsen 
Looks much simpler now.

vmassol commented:
Thanks for the PR @Haxsen!

Haxsen commented:
Thanks for the review and your time @vmassol.
I haven't updated this PR but done the requested changes except CloneCommand one.
I will update you on this in the morning :)

Haxsen committed:
getRepositoryBare - Replace authentication by CloenCommand

Haxsen committed:
Use CloneCommand object instead of separate params

Haxsen committed:
getRepositoryBare - Replace extra params by CloneCommand

Haxsen committed:
Use CloneCommand and use isBare() to test bare

Haxsen committed:
Minimize duplication and keep coverage more than 0.91

Haxsen commented:
PR updated.
Compiled and Tested on XWiki as well.

Haxsen committed:
Minimize duplication - total tests 3 - Added getRepositoryBareAndCheckBranch()

Haxsen committed:
Fix checkstyle violation: Assignment of cloneCommand is not allowed

Haxsen commented:
PR updated 
- Fixed tests duplication.
- Fixed checkstyle violation in `DefaultGitManager` (assigning of cloneCommand).

vmassol commented on src/main/java/org/xwiki/git/GitManager.java:
```diff
@@ -68,6 +69,23 @@ default Repository getRepository(String repositoryURI, String localDirectoryName
+     * @since 9.10
+     */
+    @Unstable
+    default Repository getRepositoryBare(String repositoryURI, String localDirectoryName, CloneCommand cloneCommand)
```
I would not name it `getRepositoryBare`  but `getRepository`  so that it's generic. It's also not right to both pass a CloneCommand and set it in the code. The passed CloneCommand is supposed to be readonly (final).

Could you please change that? f you really think it's good to have a `getRepositoryBare` API, then the signature should be `getRepositoryBare(String repositoryURI, String localDirectoryName)` + `getRepositoryBare(String repositoryURI, String localDirectoryName, String username, String accessCode)`.

Haxsen replied:
Yes `getRepository` is better here. We also won't need the private method (since it will be public method which can be used by other overloads in return).

vmassol commented on src/main/java/org/xwiki/git/script/GitScriptService.java:
```diff
@@ -99,6 +102,44 @@ public Repository getRepository(String repositoryURI, String localDirectoryName,
+        cloneCommand.setCredentialsProvider(new UsernamePasswordCredentialsProvider(username, accessCode));
+        return this.gitManager.getRepositoryBare(repositoryURI, localDirectoryName, cloneCommand);
+    }
+
```
You're missing the generic `getRepository` signature accepting a `CloneCommand`. Right now you cannot call it from a script which is not good. Note that this since it's not possible to create objects from Velocity, we probably need a `createCloneCommand()` method too.

Haxsen replied:
I don't think we should add the `CloneCommand` here *atm*. It would be good yes, but referring to this specific issue (bare cloning), we should add `getRepositoryBare` and later on we can add `getRepository` with the `CloneCommand`. I mean, if someone wants only the bare use case, he won't have to go through `CloneCommand`.
What do you say?

vmassol replied:
I say that bare cloning and creds are not the only use cases, so I would still add a `createCloneCommand()`  + the getRepo siganture acccepting a CloneCommand.

Haxsen replied:
Ok, I have done all the changes except this one which is what I am going to do today.
How should I make the `createCloneCommand()`. Like `createCloneCommand(setBare, creds...)` or can  you suggest a better way? 
Thanks.

vmassol replied:
`createCloneCommand()`.

Haxsen replied:
Got it.

Haxsen replied:
@vmassol  I have commited changes.
For this specifically: I have created `createCloneCommand()` in GitScriptService. I can create it in the `GitManager` interface and implement it in `DefaultGitManager` and use that one instead, but I am unsure if I should.

vmassol commented on src/test/java/org/xwiki/git/script/GitScriptServiceTest.java:
```diff
@@ -81,37 +79,37 @@ public void setupRepository() throws Exception
         GitScriptService service = this.componentManager.getInstance(ScriptService.class, "git");
         Repository repository = service.getRepository(this.testRepository.getAbsolutePath(), TEST_REPO_CLONED,
-                "test author", "TestAccessCode");
+            "test author", "TestAccessCode");
```
why use credentials for a `getRepositoryAndCountCommits`  test?

vmassol replied:
You need a test to test credentials.

Haxsen replied:
I have removed credentials from all of `GitScriptServiceTest` since the credentials are tested by `DefaultGitManagerTest`. Let me know if this is not good.
Edit: The coverage is going low since the credentials method is not covered. Adding it.

vmassol commented on src/test/java/org/xwiki/git/script/GitScriptServiceTest.java:
```diff
@@ -81,37 +79,37 @@ public void setupRepository() throws Exception
+    }
+
+    @Test
+    public void getRepositoryBareAndCheckBranch() throws Exception
```
The naming is not quite correct. It needs to represent the fixture and not the asserts, so you should remove "AndCheckBranch".

Haxsen replied:
Got it.

vmassol commented on src/test/java/org/xwiki/git/script/GitScriptServiceTest.java:
```diff
@@ -81,37 +79,37 @@ public void setupRepository() throws Exception
+    {
+        GitScriptService service = this.componentManager.getInstance(ScriptService.class, "git");
+        Repository repository = service.getRepositoryBare(this.testRepository.getAbsolutePath(), TEST_REPO_CLONED,
+            "test author", "TestAccessCode");
```
Do you have a bare cloning test without credentials?

vmassol replied:
That should be in the test name!

vmassol commented:
Waiting for new commits before reviewing again.

Haxsen committed:
Add `getRepository` method with CloneCommand

Also combines the older two methods with a return.

Haxsen committed:
Add `getRepository()` method with CloneCommand param

Haxsen committed:
Add `createCloneCommand()` , `getRepository()` with CloneCommand

Haxsen committed:
Test new `getRepository` method with CloneCommand

Use case is: `.setBare(true)`.

Haxsen committed:
Fix with/without credentials , add bare test 

Bare Test: (for createCloneCommand and getRepository method with cloneCommand param).

Haxsen commented on src/main/java/org/xwiki/git/internal/DefaultGitManager.java:
```diff
@@ -70,13 +70,7 @@
-
-    @Override
-    public Repository getRepository(String repositoryURI, String localDirectoryName, String username, String accessCode)
+    public Repository getRepository(String repositoryURI, String localDirectoryName, CloneCommand cloneCommand)
```
I can do `getRepository(CloneCommand cloneCommand)` in interface and implement here and create private method for localDirectory and builder. I am unsure about this too. Let me know if I should do  `getRepository(CloneCommand cloneCommand)`.

Haxsen commented on src/main/java/org/xwiki/git/internal/DefaultGitManager.java:
```diff
@@ -112,6 +102,23 @@ public Repository getRepository(String repositoryURI, String localDirectoryName,
+    public Repository getRepository(String repositoryURI, String localDirectoryName)
+    {
+        // Simple CloneCommand with default settings
+        CloneCommand cloneCommand = Git.cloneRepository();
```
Here we can do `createCloneCommand()` by adding it in the interface (so the `ScriptService` would also use it instead).

vmassol commented:
@Haxsen Thanks! At last I was able to push it :)

Next time,  please always use the jira issue and description in all commits. See https://dev.xwiki.org/xwiki/bin/view/Community/DevelopmentPractices#HRule:AlwaysputaJIRAissuereferenceincommitmessages

vmassol commented:
Now we need documentation. Could you please update https://extensions.xwiki.org/xwiki/bin/view/Extension/Git%20Module ?

Thanks

vmassol commented:
@Haxsen there's another problem. The tests you added are incorrectly saving data in the source tree, that's bad! Please use the `TemporaryFolder` rule.

Haxsen commented:
Thanks @vmassol . Sure, let me update the documentation and then use the `TemporaryFolder` rule.
